Ensuring that JsonAdapter annotation works correctly for primitive fi…

inder123 opened the pull request:
…elds.

inder123 commented:
Fix for https://github.com/google/gson/issues/831

swankjesse commented on gson/src/main/java/com/google/gson/internal/bind/TypeAdapterRuntimeTypeWrapper.java:
```diff
@@ -51,7 +52,7 @@ public void write(JsonWriter out, T value) throws IOException {
     TypeAdapter chosen = delegate;
     Type runtimeType = getRuntimeTypeIfMoreSpecific(type, value);
-    if (runtimeType != type) {
+    if (runtimeType != type && !Primitives.areEquivalentTypes(runtimeType, type)) {
```
This doesn’t seem like the right location for the fix. I’d expect it to be nearby the `@JsonAdapter` handling code.

inder123 replied:
There is no other way that I can think of. The problem is that ReflectiveTypeAdapterFactory (line 109) field.get(value) returns Long even though the field is of type long. This results in the runtime type being incorrectly characterized for all primitive types.  Hence the TypeAdapterRuntimeTypeWrapper needs to be smarter when dealing with primitive wrapper types.

swankjesse replied:
The problem is triggered by primitive vs. wrapper types, but that isn't the root of the problem. The root of the problem is that Gson currently lets the runtime type take precedence over the `@JsonAdapter` annotation. This is weird!

For example, consider a JSON adapter that calls `toString()` on the object and writes that string to the `JsonWriter`:

```
  private static class ToStringAdapter extends TypeAdapter<Object> {
    @Override public Object read(JsonReader in) throws IOException {
      throw new UnsupportedOperationException();
    }

    @Override public void write(JsonWriter out, Object value) throws IOException {
      out.value(value.toString());
    }
  }
```

We can request this adapter like so:

```
  private static final class Box {
    @JsonAdapter(ToStringAdapter.class)
    Object o;
  }
```

But unfortunately due to a mistake in how Gson precedence works, the `ToStringAdapter` is ignored whenever `o`’s type isn’t `Object.class`. For example, this test fails because the `ToStringAdapter` is never used.

```
  public void test() {
    Gson gson = new Gson();

    Box box = new Box();
    box.o = Arrays.asList(5, 6, 7);

    String json = gson.toJson(box);
    assertEquals("{\"o\":\"" + box.o.toString() + "\"}", json);
  }
```

Fixing the precedence is [possible](https://github.com/google/gson/commit/d1770c9f77cd7761add943d81b773b2df4594f3b), but dangerous and potentially incompatible with currently deployed code.

inder123 replied:
I see your point. I think we should fix the precedence for JsonAdapter as that is the expected behavior by clients. I think JsonAdapter is new enough to not have serious incompatibility issues (as we discovered with this primitive-related behavior).
I will update the pull request.

JakeWharton replied:
We've fixed precedence before: 7f8f490fdc95fb087e735dc582a2af07cee46078

swankjesse replied:
Sounds good!

swankjesse replied:
Is this change still necessary?

inder123 commented:
@swankjesse I have revised the code per feedback. Now if JsonAdapter annotation is present, we ignore  the run-time type specific adapter.

Also, rebased and squashed. Please review.

swankjesse commented on gson/src/test/java/com/google/gson/functional/JsonAdapterAnnotationOnFieldsTest.java:
```diff
@@ -220,4 +220,44 @@ private GadgetWithOptionalPart(Part part) {
   }
+
+  /** Regression test contributed through https://github.com/google/gson/issues/831 */
+  public void testPrimitiveFieldAnnotationTakesPrecedenceOverDefault() {
```
Worth having a test for the non-primitive case?

inder123 replied:
Yes, was being lazy.. :)

swankjesse commented on gson/src/main/java/com/google/gson/internal/bind/TypeAdapterRuntimeTypeWrapper.java:
```diff
@@ -51,7 +52,7 @@ public void write(JsonWriter out, T value) throws IOException {
     Type runtimeType = getRuntimeTypeIfMoreSpecific(type, value);
-    if (runtimeType != type) {
+    if (runtimeType != type && !Primitives.areEquivalentTypes(runtimeType, type)) {
       TypeAdapter runtimeTypeAdapter = context.getAdapter(TypeToken.get(runtimeType));
```
If it is, in theory it should just happen in here

inder123 replied:
Good point. Will make that change.

inder123 replied:
On second thoughts, I don't really see an advantage of moving the code here. Can you explain what you mean by providing a snippet?

swankjesse replied:
What I intended to say was that this line will do the primitive/wrapper work automatically so the check above is unnecessary.

inder123 commented:
@swankjesse  Ok. I have removed that check. Please review.

swankjesse commented:
Thumbs up
